File: src/main/scala/com/spr/scala/syntax/AnonymousFunctions.scala
package com.spr.scala.syntax

/**
  * Demonstrates various ways to create anonymous functions.
  */
object AnonymousFunctions {

  def anonymousClassSyntax: String => Int = new Function[String, Int] {
    override def apply(v1: String): Int = Integer.parseInt(v1)
  }

  def lambdaSyntax: String => Int = s => Integer.parseInt(s)

  def altLambdaSyntax: String => Int = { s =>
    Integer.parseInt(s)
  }

  def lambdaAnonymousParameterSyntax: String => Int = Integer.parseInt(_)

  def functionReferenceSyntax: String => Int = Integer.parseInt

  def multiParamLambdaSyntax: (String, Int) => Boolean = (s, i) => Integer.parseInt(s) == i
}

File: src/main/java/com/spr/java/syntax/AnonymousFunctions.java
package com.spr.java.syntax;

import java.util.function.BiFunction;
import java.util.function.Function;

/**
 * Demonstrates various ways to create anonymous functions.
 */
public class AnonymousFunctions {

    public static Function<String, Integer> anonymousClassSyntax() {
        return new Function<String, Integer>() {
            @Override
            public Integer apply(final String s) {
                return Integer.parseInt(s);
            }
        };
    }

    public static Function<String, Integer> lambdaSyntax() {
        return s -> Integer.parseInt(s);
    }

    public static Function<String, Integer> altLambdaSyntax() {
        return s -> {
            return Integer.parseInt(s);
        };
    }

    public static Function<String, Integer> functionReferenceSyntax() {
        return Integer::parseInt;
    }

    public static BiFunction<String, Integer, Boolean> multiParamLambdaSyntax() {
        return (s, i) -> Integer.parseInt(s) == i;
    }
}
